File: common/src/main/java/com/hartwig/actin/molecular/filter/GeneAggregator.java
package com.hartwig.actin.molecular.filter;

import java.util.Collection;
import java.util.stream.Collectors;

import com.hartwig.serve.datamodel.common.GeneRole;
import com.hartwig.serve.datamodel.gene.KnownGene;

import org.jetbrains.annotations.NotNull;

public final class GeneAggregator {

    public static Collection<KnownGene> aggregate(@NotNull Collection<KnownGene> rawGenes) {
        return rawGenes.stream()
                .collect(Collectors.groupingBy(KnownGene::gene))
                .values()
                .stream()
                .map(genes -> genes.stream()
                        .filter(gene -> !gene.geneRole().equals(GeneRole.UNKNOWN))
                        .findFirst()
                        .orElse(genes.iterator().next()))
                .collect(Collectors.toSet());
    }
}

File: common/src/main/kotlin/com/hartwig/actin/molecular/evidence/known/GeneAggregator.kt
package com.hartwig.actin.molecular.evidence.known

import com.hartwig.serve.datamodel.molecular.common.GeneRole
import com.hartwig.serve.datamodel.molecular.gene.KnownGene

internal object GeneAggregator {

    private val ROLE_PRECEDENCE = listOf(GeneRole.BOTH, GeneRole.ONCO, GeneRole.TSG, GeneRole.UNKNOWN)

    fun aggregate(rawGenes: Collection<KnownGene>): Set<KnownGene> {
        return rawGenes.groupBy(KnownGene::gene).values
            .map { genes -> genes.minWith(genePrecedenceOrder()) }
            .toSet()
    }

    private fun genePrecedenceOrder(): Comparator<KnownGene> {
        return Comparator.comparingInt { ROLE_PRECEDENCE.indexOf(it.geneRole()) }
    }
}
